File: defence_claw/defenseclaw/cli/defenseclaw/ux.py
# Copyright 2026 Cisco Systems, Inc. and its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Small terminal renderer shared by first-run commands.

The renderer intentionally keeps presentation out of the bootstrap
backend. It honors non-TTY/NO_COLOR output, gives operators concise step
lines, and still produces plain text that is friendly to CI logs.

In addition to the :class:`CLIRenderer` (used by the structured
first-run pipeline), this module exports a handful of small *free
functions* — :func:`section`, :func:`subhead`, :func:`ok`,
:func:`warn`, :func:`err`, :func:`hint`, :func:`kv`, :func:`bold`,
:func:`dim`, and :func:`accent` — so ad-hoc setup wizards (e.g.
``defenseclaw setup guardrail``) can colorize their output without
instantiating a renderer per call site. They share the same TTY +
``NO_COLOR`` detection rule as :class:`CLIRenderer`, but the gate is
recomputed on every call so a test that monkey-patches
``sys.stdout`` or ``os.environ`` behaves predictably.
"""

from __future__ import annotations

import os
import sys
from dataclasses import dataclass

import click

TUI_UNAVAILABLE_MESSAGE = (
    "The interactive dashboard needs a UTF-8 capable terminal. "
    "On Windows, open Windows Terminal or PowerShell 7 and run 'defenseclaw' again. "
    "This terminal can still run 'defenseclaw status' and 'defenseclaw doctor'."
)

# ``main()`` snapshots this before it reconfigures Python's streams to UTF-8.
# Without the snapshot, a legacy cp1252/OEM stream would look capable after the
# reconfigure even though its terminal host still cannot render these glyphs.
_configured_unicode_output: bool | None = None

_UNICODE_PROBE = "✓✗⚠─═└—↪"
_ASCII_PRESENTATION_TRANSLATION = str.maketrans(
    {
        "✓": "OK",
        "✔": "OK",
        "✗": "X",
        "✘": "X",
        "⚠": "!",
        "─": "-",
        "━": "-",
        "═": "=",
        "│": "|",
        "║": "|",
        "┌": "+",
        "┐": "+",
        "└": "\\",
        "┘": "+",
        "├": "+",
        "┤": "+",
        "┬": "+",
        "┴": "+",
        "┼": "+",
        "╭": "+",
        "╮": "+",
        "╯": "+",
        "╰": "+",
        "—": "-",
        "–": "-",
        "→": "->",
        "←": "<-",
        "↪": "->",
        "…": "...",
        "•": "*",
        "·": "-",
        "●": "*",
        "○": "o",
    }
)


def _stream_is_tty(stream: object) -> bool:
    try:
        return bool(stream.isatty())  # type: ignore[attr-defined]
    except (AttributeError, OSError, ValueError):
        return False


def _stream_supports_unicode(stream: object) -> bool:
    """Return whether an interactive stream can encode CLI presentation glyphs."""

    if not _stream_is_tty(stream):
        # Redirected output is deliberately stable ASCII even when the target
        # file happens to use UTF-8. Machine-readable JSON bypasses this layer.
        return False
    try:
        encoding = getattr(stream, "encoding", None)
    except (AttributeError, OSError, ValueError):
        return False
    if not isinstance(encoding, str) or not encoding:
        return False
    try:
        _UNICODE_PROBE.encode(encoding)
    except (LookupError, UnicodeEncodeError):
        return False
    return True


def configure_console_output(stream: object | None = None) -> bool:
    """Snapshot whether human CLI output may use rich Unicode presentation.

    ``TERM=dumb`` is the explicit native-launch fallback. Otherwise stdout must
    be an interactive stream whose original encoding supports the complete
    presentation-glyph set. Call this before any UTF-8 stream reconfiguration.
    """

    global _configured_unicode_output
    target = sys.stdout if stream is None else stream
    _configured_unicode_output = os.environ.get("TERM", "").strip().lower() != "dumb" and _stream_supports_unicode(
        target
    )
    return _configured_unicode_output


def unicode_output_enabled() -> bool:
    """Return the policy snapshot, preserving rich output before configuration.

    The default keeps direct library/Click-test use backward-compatible. The
    shipped entrypoint always calls :func:`configure_console_output` first.
    """

    if os.environ.get("TERM", "").strip().lower() == "dumb":
        return False
    if _configured_unicode_output is None:
        return True
    return _configured_unicode_output


def console_text(text: str) -> str:
    """Downgrade presentation glyphs while preserving ordinary Unicode text."""

    if unicode_output_enabled():
        return text
    return text.translate(_ASCII_PRESENTATION_TRANSLATION)


def echo(message: object | None = None, **kwargs: object) -> None:
    """Call :func:`click.echo` with presentation-safe human output."""

    if isinstance(message, str):
        message = console_text(message)
    click.echo(message, **kwargs)


def terminal_supports_tui(*, stdin: object | None = None, stdout: object | None = None) -> bool:
    """Return whether both streams and the rendering policy can host Textual."""

    input_stream = sys.stdin if stdin is None else stdin
    output_stream = sys.stdout if stdout is None else stdout
    return unicode_output_enabled() and _stream_is_tty(input_stream) and _stream_is_tty(output_stream)


@dataclass
class CLIRenderer:
    """Minimal status renderer for CLI setup flows."""

    color: bool | None = None
    quiet: bool = False

    def __post_init__(self) -> None:
        if self.color is None:
            self.color = unicode_output_enabled() and sys.stdout.isatty() and "NO_COLOR" not in os.environ

    def echo(self, text: str = "", *, err: bool = False) -> None:
        if self.quiet:
            return
        echo(text, err=err)

    def title(self, text: str, subtitle: str = "") -> None:
        if self.quiet:
            return
        self.echo()
        self.echo(self._style(f"  {text}", fg="cyan", bold=True))
        if subtitle:
            self.echo(self._style(f"  {subtitle}", fg="bright_black"))
        self.echo("  " + self._style("─" * 56, fg="bright_black"))

    def section(self, text: str) -> None:
        if self.quiet:
            return
        self.echo()
        self.echo(self._style(f"  {text}", fg="bright_black", bold=True))

    def step(self, status: str, label: str, detail: str = "") -> None:
        if self.quiet:
            return
        icon = {
            "pass": "✓",
            "warn": "!",
            "fail": "x",
            "skip": "-",
        }.get(status, "-")
        fg = {
            "pass": "green",
            "warn": "yellow",
            "fail": "red",
            "skip": "bright_black",
        }.get(status, "white")
        line = f"  {self._style(icon, fg=fg, bold=True)} {label}"
        if detail:
            line += self._style(f"  {detail}", fg="bright_black")
        self.echo(line)

    def _style(self, text: str, **kwargs) -> str:
        text = console_text(text)
        if not self.color:
            return text
        return click.style(text, **kwargs)


# ---------------------------------------------------------------------------
# Module-level free helpers used by ad-hoc setup wizards
# ---------------------------------------------------------------------------
#
# Design contract (shared by every helper below):
#
#   * Each helper reads :func:`_color_enabled` per call so tests that
#     monkey-patch ``sys.stdout`` or ``NO_COLOR`` see the new state
#     immediately. Caching the gate on import would freeze the
#     decision to whatever the first call observed — that bit a
#     past iteration of this module so the live evaluation is
#     deliberate.
#   * Honors the de-facto cross-tool conventions:
#       - ``NO_COLOR`` env var (any value) disables colors.
#         See https://no-color.org for the cross-vendor spec.
#       - Non-TTY stdout disables colors.
#       - ``CLICOLOR_FORCE`` / ``FORCE_COLOR`` can opt back into color only
#         when the snapshotted rendering policy is Unicode-capable.
#   * Returns plain strings (no side-effects) so callers can compose
#     them inside ``f"…"`` and ``click.echo``. A separate helper
#     (:func:`section`) emits two lines for headings; that one prints
#     directly because the divider is ALWAYS bound to the heading.
#
# These helpers exist BECAUSE :class:`CLIRenderer` is overkill for
# wizard flows that already do their own ``click.echo`` layout. They
# are intentionally additive: existing call sites that use
# :class:`CLIRenderer` keep working unchanged.


def _color_enabled() -> bool:
    """Return ``True`` when colorized output is appropriate.

    Recomputed on every call so monkey-patched stdout / env vars in
    tests take effect immediately. Order:

      1. ASCII-safe / ``TERM=dumb`` rendering → False. ANSI controls are not
         safe to inject into the final legacy fallback.
      2. ``CLICOLOR_FORCE`` / ``FORCE_COLOR`` truthy → True (force
         colors even when not a TTY; the standard "yes I really
         want colors in my piped log" escape hatch).
      3. ``NO_COLOR`` set (any value) → False. Per
         https://no-color.org any presence — including empty — means
         "no color".
      4. ``sys.stdout.isatty()`` → use that.
    """
    if not unicode_output_enabled():
        return False
    if os.environ.get("CLICOLOR_FORCE", "").strip() or os.environ.get("FORCE_COLOR", "").strip():
        return True
    if "NO_COLOR" in os.environ:
        return False
    try:
        return bool(sys.stdout.isatty())
    except (AttributeError, ValueError):
        # ``sys.stdout`` may have been swapped for a non-stream
        # object in tests or under unusual reentrancy; degrade to
        # "no color" rather than crashing.
        return False


def _style(text: str, **kwargs: object) -> str:
    """Wrap :func:`click.style` with the color gate.

    Centralized so every helper picks up policy changes (NO_COLOR,
    forced color, TTY downgrade) from a single check.
    """
    text = console_text(text)
    if not _color_enabled():
        return text
    return click.style(text, **kwargs)


def bold(text: str) -> str:
    """Bold the text (no color change)."""
    return _style(text, bold=True)


def dim(text: str) -> str:
    """Dim text — typically for explanatory copy below a heading."""
    return _style(text, fg="bright_black")


def accent(text: str) -> str:
    """Cyan accent for inline emphasis on key concepts."""
    return _style(text, fg="cyan")


def hint(text: str) -> str:
    """Dim hint text — same color as :func:`dim` but spelled out
    for caller intent (``hint('...')`` reads better than ``dim``
    when the line is a parenthetical aside)."""
    return _style(text, fg="bright_black")


def section(title: str, *, indent: str = "  ", divider_char: str = "─") -> None:
    """Print a bold cyan heading with a colored divider underneath.

    Two lines are emitted — heading and divider — and a leading
    blank line precedes the heading so back-to-back sections don't
    visually run together. Divider length matches ``len(title)`` so
    the underline tracks the heading width.

    Use for wizard-level section breaks (``LLM Guardrail Setup``,
    ``Hook fail mode``, ``Human Approval``, etc.). For inline
    emphasis use :func:`accent` instead.
    """
    echo()
    echo(f"{indent}{_style(title, fg='cyan', bold=True)}")
    echo(f"{indent}{_style(divider_char * len(title), fg='cyan')}")


def banner(
    title: str,
    *,
    indent: str = "  ",
    width: int = 54,
    divider_char: str = "─",
    leading_blank: bool = True,
) -> None:
    """Print a full-width ``── Title ─────…──`` banner.

    Used by long-form flows (``defenseclaw init``,
    ``defenseclaw upgrade``, ``defenseclaw uninstall``) where the
    section dividers are tall and wide so the operator can locate
    them when scrolling back through a 200-line transcript.

    The banner format is ``── <title> ───…───`` extending to ``width``
    columns; we intentionally keep the layout legacy-compatible
    when colors are off so existing tests that grep for
    ``"── Environment ──"`` keep matching unchanged. Color-on
    bolds the title and dims the dashes so the title pops without
    making the dashes shouty.
    """
    if leading_blank:
        echo()
    label = f" {title} "
    side = max(2, (width - len(label)) // 2)
    left = divider_char * side
    right = divider_char * (width - side - len(label))
    if _color_enabled():
        echo(
            f"{indent}{_style(left, fg='bright_black')}"
            f" {_style(title, fg='cyan', bold=True)} "
            f"{_style(right, fg='bright_black')}"
        )
    else:
        # Plain mode keeps the legacy "── Title ──────...──"
        # format exactly so any test or screen scraper that
        # grep-substrings on ``"── Environment ──"`` keeps
        # matching.
        echo(f"{indent}{left}{label}{right}")
    echo()


def subhead(text: str, *, indent: str = "  ") -> None:
    """Print a single dim subhead/explanatory line.

    Mirrors :meth:`CLIRenderer.section` color — bright_black so it
    visually recedes below the cyan heading without disappearing.
    """
    echo(f"{indent}{dim(text)}")


def ok(text: str, *, indent: str = "  ", marker: str = "✓") -> None:
    """Print a green success line: ``  ✓ {text}``."""
    echo(f"{indent}{_style(marker, fg='green', bold=True)} {text}")


def warn(text: str, *, indent: str = "  ", marker: str = "⚠") -> None:
    """Print a yellow warning line: ``  ⚠ {text}``.

    Use for non-fatal advisories (e.g., "Configuration not saved";
    "redaction is OFF in shared deployments"). Reserve :func:`err`
    for genuinely failed operations.
    """
    echo(f"{indent}{_style(marker, fg='yellow', bold=True)} {_style(text, fg='yellow')}")


def err(text: str, *, indent: str = "  ", marker: str = "✗") -> None:
    """Print a red error line: ``  ✗ {text}``.

    Output goes to stdout (not stderr) because Click's ``echo``
    convention in this codebase is to mix all wizard output on the
    same channel for screen-reader and copy-paste predictability.
    Callers that genuinely need stderr should use
    :func:`click.echo` with ``err=True`` directly.
    """
    echo(f"{indent}{_style(marker, fg='red', bold=True)} {_style(text, fg='red')}")


def kv(
    key: str,
    value: object,
    *,
    indent: str = "    ",
    key_width: int = 30,
) -> None:
    """Print a colored key/value row used in wizard summaries.

    The key is rendered dim+bold and right-padded so the colon
    column lines up across rows. ``key_width`` is the total
    column width (including the trailing ``":"``) — matches the
    pre-existing ``f"{key + ':':<30s} {val}"`` format that the
    guardrail summary uses, so this helper is a drop-in upgrade
    that does not shift the layout.

    The value is rendered in the default foreground so it pops
    out against the dim key. Empty / falsy values render as a
    dim em-dash so the row still occupies its column instead of
    looking truncated.
    """
    text_value = "" if value is None else str(value)
    rendered_value = dim("—") if not text_value else text_value
    label = (key + ":").ljust(key_width)
    echo(f"{indent}{_style(label, fg='bright_black', bold=True)} {rendered_value}")

File: defence_claw/defenseclaw/internal/cli/ux.go
// Copyright 2026 Cisco Systems, Inc. and its affiliates
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// SPDX-License-Identifier: Apache-2.0

package cli

import (
	"fmt"
	"os"
	"strings"
	"unicode/utf8"

	"golang.org/x/term"
)

// ANSI SGR fragments (reset applied once at end of Style).
const (
	ansiBold          = "\x1b[1m"
	ansiFgCyan        = "\x1b[36m"
	ansiFgGreen       = "\x1b[32m"
	ansiFgYellow      = "\x1b[33m"
	ansiFgRed         = "\x1b[31m"
	ansiFgBrightBlack = "\x1b[90m"
	ansiReset         = "\x1b[0m"
)

func envTruthy(key string) bool {
	return strings.TrimSpace(os.Getenv(key)) != ""
}

// ColorEnabled reports whether ANSI styling should be applied.
// Evaluated on each call (no caching) so tests and subprocesses that
// mutate the environment see updates immediately.
//
// Order matches cli/defenseclaw/ux.py::_color_enabled:
//  1. CLICOLOR_FORCE or FORCE_COLOR truthy → true (even when stdout is not a TTY).
//  2. NO_COLOR present in the environment (any value, including empty) → false.
//  3. Otherwise → true iff stdout is a terminal.
func ColorEnabled() bool {
	if envTruthy("CLICOLOR_FORCE") || envTruthy("FORCE_COLOR") {
		return true
	}
	if _, ok := os.LookupEnv("NO_COLOR"); ok {
		return false
	}
	fd := int(os.Stdout.Fd())
	return term.IsTerminal(fd)
}

// Style applies zero or more attributes then resets.
// Supported attrs: "bold", "fg=cyan", "fg=green", "fg=yellow", "fg=red", "fg=bright_black".
// When colors are disabled, text is returned unchanged.
func Style(text string, attrs ...string) string {
	if !ColorEnabled() || len(attrs) == 0 {
		return text
	}
	var b strings.Builder
	for _, a := range attrs {
		switch a {
		case "bold":
			b.WriteString(ansiBold)
		default:
			if after, ok := strings.CutPrefix(a, "fg="); ok {
				switch after {
				case "cyan":
					b.WriteString(ansiFgCyan)
				case "green":
					b.WriteString(ansiFgGreen)
				case "yellow":
					b.WriteString(ansiFgYellow)
				case "red":
					b.WriteString(ansiFgRed)
				case "bright_black":
					b.WriteString(ansiFgBrightBlack)
				}
			}
		}
	}
	b.WriteString(text)
	b.WriteString(ansiReset)
	return b.String()
}

// Bold wraps text in bold SGR (no color change).
func Bold(text string) string {
	return Style(text, "bold")
}

// Dim renders text in bright black / gray.
func Dim(text string) string {
	return Style(text, "fg=bright_black")
}

// Accent renders cyan emphasis for inline highlights.
func Accent(text string) string {
	return Style(text, "fg=cyan")
}

// Section prints a bold cyan heading with a cyan underline matching the title width.
func Section(title string) {
	fmt.Println()
	fmt.Println("  " + Style(title, "fg=cyan", "bold"))
	w := utf8.RuneCountInString(title)
	fmt.Println("  " + Style(strings.Repeat("─", w), "fg=cyan"))
}

// Banner prints a full-width "── Title ──…──" line (~54 cols) plus a trailing blank line.
func Banner(title string) {
	fmt.Println()
	const width = 54
	const indent = "  "
	label := fmt.Sprintf(" %s ", title)
	side := max(2, (width-len(label))/2)
	left := strings.Repeat("─", side)
	right := strings.Repeat("─", width-side-len(label))
	if ColorEnabled() {
		fmt.Printf("%s%s %s %s\n", indent, Dim(left), Style(title, "fg=cyan", "bold"), Dim(right))
	} else {
		fmt.Printf("%s%s%s%s\n", indent, left, label, right)
	}
	fmt.Println()
}

// Subhead prints one dim explanatory line.
func Subhead(text string) {
	fmt.Println("  " + Dim(text))
}

// OK prints a green success marker line.
func OK(text string) {
	fmt.Printf("  %s %s\n", Style("✓", "fg=green", "bold"), text)
}

// Warn prints a yellow warning marker line.
func Warn(text string) {
	fmt.Printf("  %s %s\n", Style("⚠", "fg=yellow", "bold"), Style(text, "fg=yellow"))
}

// Err prints a red error marker line (stdout; matches Python ux.err convention).
func Err(text string) {
	fmt.Printf("  %s %s\n", Style("✗", "fg=red", "bold"), Style(text, "fg=red"))
}

// KV prints a dim bold key column and a plain value (30-char label column, 4-space indent).
func KV(key string, value any) {
	const indent = "    "
	const keyWidth = 30
	textValue := ""
	if value != nil {
		textValue = fmt.Sprint(value)
	}
	rendered := textValue
	if rendered == "" {
		rendered = Dim("—")
	}
	label := fmt.Sprintf("%-*s", keyWidth, key+":")
	fmt.Printf("%s%s %s\n", indent, Style(label, "fg=bright_black", "bold"), rendered)
}
